src/colmap/estimators/pose.cc

```diff
@@ -84,83 +51,46 @@ bool EstimateAbsolutePose(const AbsolutePoseEstimationOptions& options,
-    // Generate focal length factors using a quadratic function,
-    // such that more samples are drawn for small focal lengths
-    focal_length_factors.reserve(options.num_focal_length_samples + 1);
-    const double fstep = 1.0 / options.num_focal_length_samples;
-    const double fscale =
-        options.max_focal_length_ratio - options.min_focal_length_ratio;
-    double focal = 0.;
-    for (size_t i = 0; i <= options.num_focal_length_samples;
-         ++i, focal += fstep) {
-      focal_length_factors.push_back(options.min_focal_length_ratio +
-                                     fscale * focal * focal);
-    }
-  } else {
-    focal_length_factors.reserve(1);
-    focal_length_factors.push_back(1);
-  }
-
-  std::vector<std::future<void>> futures;
-  futures.resize(focal_length_factors.size());
-  std::vector<typename AbsolutePoseRANSAC::Report,
-              Eigen::aligned_allocator<typename AbsolutePoseRANSAC::Report>>
-      reports;
-  reports.resize(focal_length_factors.size());
-
-  ThreadPool thread_pool(std::min(
-      options.num_threads, static_cast<int>(focal_length_factors.size())));
-
-  for (size_t i = 0; i < focal_length_factors.size(); ++i) {
-    futures[i] = thread_pool.AddTask(EstimateAbsolutePoseKernel,
-                                     *camera,
-                                     focal_length_factors[i],
-                                     points2D,
-                                     points3D,
-                                     options.ransac_options,
-                                     &reports[i]);
-  }
-
-  double focal_length_factor = 0;
-  Eigen::Matrix3x4d cam_from_world_matrix;
   *num_inliers = 0;
   inlier_mask->clear();
 
-  // Find best model among all focal lengths.
-  for (size_t i = 0; i < focal_length_factors.size(); ++i) {
-    futures[i].get();
-    const auto report = reports[i];
-    if (report.success && report.support.num_inliers > *num_inliers) {
-      *num_inliers = report.support.num_inliers;
-      cam_from_world_matrix = report.model;
-      *inlier_mask = report.inlier_mask;
-      focal_length_factor = focal_length_factors[i];
-    }
+  std::vector<Eigen::Vector2d> points2D_normalized(points2D.size());
+  for (size_t i = 0; i < points2D.size(); ++i) {
+    points2D_normalized[i] = camera->CamFromImg(points2D[i]);
   }
 
-  if (*num_inliers == 0) {
-    return false;
-  }
+  auto custom_ransac_options = options.ransac_options;
+  custom_ransac_options.max_error =
+      camera->CamFromImgThreshold(options.ransac_options.max_error);
 
-  // Scale output camera with best estimated focal length.
-  if (options.estimate_focal_length && *num_inliers > 0) {
-    for (const size_t idx : camera->FocalLengthIdxs()) {
-      camera->params[idx] *= focal_length_factor;
+  if (options.estimate_focal_length) {
+    // TODO(joschonb): Implement non-minimal solver for LORANSAC refinement.
+    RANSAC<P4PFEstimator> ransac(custom_ransac_options);
```

Comment:
This should always be called by the RefineAbsolutePose, which refines on all inliers. As I understood your earlier experiments, this yielded essentially identical results to using LORANSAC?

Reply:
In the initial experiments on ArtsQuad instances this seemed to be the case, but in your initial tests you had failures there and I am not sure how this would be different. I did not run this exact comparison on the larger dataset (from MegaScenes). I can quickly see if I can get these numbers.

Reply:
Thank you for the experiment. Above, I thought that poselib-pnpf+RAP vs colmap-EAP+RAP would be essentially a comparison of this branch vs main. Probably, I misunderstood. Let's see your numbers to be sure.

Reply:
Ah yes, so poselib-pnpf is with LO-RANSAC.

Here are the results without LO in the RANSAC. So poselib-pnpf and poselib-pnpf-NO_LO are exactly the same exact in each call for LO it just returns without doing anything.

<img width="1205" alt="image" src="https://github.com/user-attachments/assets/52a7e3b0-c6ac-40e8-928c-d770b32d8969">

I think the ransac parameters might've been different in this run from the last (as they are all a bit slower), but I guess it should not matter for the comparison.

Reply:
Thank you very much. With these experiments, I conclude that we can move ahead with the current code and leave further improvements to the future. It will become easier once you merge your changes in the poselib dev branch to main. Then, we can consider directly moving to the higher-level poselib::estimate_absolute_pose functions, even though we need to be careful because of possible discrepancies of the camera models.